service/r2/kv/store.go

```diff
@@ -73,6 +75,10 @@ func (s *store) FetchNamespaces() (*rules.NamespacesView, error) {
 	}, nil
 }
 
+func (s *store) ValidateNamespace(rs *rules.RuleSetSnapshot) error {
+	return s.validator.ValidateSnapshot(rs)
```

Comment:
I'd check if `s.validator` is nil and return an `errNilValidator = errors.New("validator not provided")` just to be defensive.